_episodes/32.md

```diff
@@ -0,0 +1,389 @@
+resources at hand. We cannot change the plan for a subquery if we notice that 
+decision already made is not optimal.
+
+### Trino engine improvements with Project Tardigrade
+
+Project Tardigrade aims to break the all-or-nothing execution barrier that would
+open many new opportunities around resource management, adaptive query 
+optimization, and failure recovery. We are planning to achieve this by 
+introducing spooling using an efficient buffering layer at stage boundaries. The
+buffer stores intermediate results for the duration of a query or a stage, 
+depending on the context. The project is named after the microscopic [Tardigrades](https://en.wikipedia.org/wiki/Tardigrade)
+that are the world's most indestructible creatures, akin to the resiliency we 
+are adding to Trino.
+
+<p align="center">
+<img align="center" width="50%" height="100%" src="/assets/episode/32/tardigrade-logo.png"/><br/>
+</p>
+
+Buffering intermediate results makes it possible to execute queries iteratively.
+For example, the engine can process one or several tasks at a time effectively 
+reducing memory pressure and allow memory intensive queries to succeed without 
+a need to expand the cluster. It would significantly lower cost of operation, 
+specifically for the situation when only a small number of queries requires more
+memory than available.
+
+<p align="center">
+<img align="center" width="50%" height="100%" src="/assets/episode/32/tardigrade-buffers.png"/><br/>
+</p>
+
+#### Adaptive Planning
+
+The engine may also decide to re-optimize the query at stage boundaries. When  
+the engine buffers the intermediate data, it is possible to get better insight
+into the nature of the data as it's processed and adapt query plans accordingly.
+For example, when the cost based optimizer makes a bad decision, because of 
+incorrect statistics or estimates, it can pick the wrong type of join, or a 
+suboptimal join order. The engine can then suspend the query, re-optimize the 
+plan, and resume the query. Additionally, it may allow the engine to discover 
+skewed datasets and change query plans accordingly. This may significantly 
+improve efficiency and landing time for workloads that are `JOIN` heavy.
+
+#### Resource Management
+
+Iterative query processing would also allow us to be more flexible at resource 
+management. Resource allocation can be adjusted as the queries run. For example,
+when a cluster is idle we may allow a single query to utilize all available 
+resources on a cluster. When more workload kicks in the resource allocation for
+the initial query can be gradually reduced and available resources can be 
+granted to newly submitted workloads. With this model it is also significantly 
+easier to implement auto scaling. When workload submitted requires more 
+resources that are currently available in the cluster the engine may request 
+more nodes to be added to the cluster. Or quite the opposite, if the cluster 
```

Comment:
Or the opposite, if the cluster is underutilized, it is easier to return resources